### DISCUSSION TITLE
Using TPM handles (keys stored or created in TPM) to create tunnel by strongswan

### DISCUSSION THREAD

**Discussion by mkr1d** [strongswan/strongswan#1898] [General]
Hi All!
I am using Ubuntu 20.04 ARM64 system,hardware  TPM 2.0 device and strongswan version 5.8.2.
I was able to create a tunnel using RSA 2048 keys and certificates,which I create in Linux and store on flash. I am using an apparently outdated configuration scheme of ipsec.conf and ipsec.secrets.
I've got a few questions:
1)Is it mandatory for me to upgrade to version 5.9 in order to be able to use my TPM 2.0 (Hardware) device with strongswan ?
2)Is it possible to use current (legacy) configuration of ipsec.conf/secrets to utilize the TPM device ?If so, can you point me to examples or provide some basic usage example?
3)In case that using swanctl.conf is mandatory:
-There is example in https://docs.strongswan.org/docs/5.9/tpm/tpm2Ike.html that uses TPM handles, which point  to attestation key/certificate  (AK) .Is it mandatory to use attestation keys (AK) ,which are created in the TPM or may I use externally created keys ,which I import to TPM ?
-Continuing the subject ,is it mandatory to use certificates signed by the TPM ,as implied in this example ?

**Comment by strongX509** [comment]
1) strongSwan 5.8.2 should be ok to use with TPM 2.0. Also the tpm2-tss package version 2.3.2 preinstalled with Ubuntu 20.04 should work with the strongSwan libtpmtss library.
2) Never tried to configure access to a TPM key with ipsec.conf. I strongly doubt that it works.
3) Yes, you should switch to swanctl configuration. The TPM 2.0 specification does not allow a private key to be imported into persistent storage. You have to generate the key on-chip and export a PKCS#10 certificate request to be signed by a CA. The TPM 2.0 itself does not sign the certificate, it just signs the certificate request.

**Reply by mkr1d** [reply]
"The TPM 2.0 specification does not allow a private key to be imported into persistent storage"->Are you sure ?
If I do the following steps (RSA key):
1)Generate RSA key (for instance,by openssl)
2)Create primary key (if doesn't exist)
3)TPM2_Import (RSA key)
4)TPM2_Load(from step 3)
5)After getting a handle in step 4 ,I make the handle persistent
So after all these steps , the key generated by openssl is not in persistent storage?What does exist in that handle which was made persistent,then ?

**Reply by mkr1d** [reply]
Thank You for your answers!
Continuing the discussion from before : Does the certificate (after signing) have to be persistent in the TPM or I can store it in transient memory ,or maybe I can store it on local storage flash ?

**Comment by strongX509** [comment]
When I follow your instructions I get the following persisted key:
```
$ pki --print --type priv --keyid 0x81010004
TPM 2.0 via TSS2 v2 available
signature algorithm is NULL with ERROR hash
encryption algorithm is NULL-ERROR with 0 bits
  privkey:   RSA 2048 bits
  keyid:     d9:24:b3:ed:38:98:3f:9d:17:d4:13:5a:7a:88:71:6a:57:4c:0d:aa
  subjkey:   e6:42:17:18:65:bf:c2:ae:5f:95:ef:0c:83:18:02:ea:aa:e0:31:37
```
This key cannot be used for signing an IKEv2 AUTH payload since no signing algorithm is defined.

An X.509 certificate doesn't have to be persisted in the NV RAM of the TPM 2.0. It can be stored as an external file.

**Comment by mkr1d** [comment]
If X.509 certificate is stored as an external file, how do I configure strongswan.conf to point it to its appropriate tpm handle ?

**Reply by strongX509** [reply]
**`strongswan.conf`** is not the correct place. Use **`/etc/swanctl/swanctl.conf`** for IPsec connection definitions.

In **`swanctl.conf`** define:
```
      ...
      local {
        auth = pubkey
        certs = myCert.pem
      }
     ...
```
for the external certificate file  **`myCert.pem`** stored in the  **`/etc/swanctl/x509/`** directory and
```
secrets {
    token_rsa {
       handle = 0x81010004
    }
}
```
for the private key stored in the TPM 2.0.

**Reply by mkr1d** [reply]
Yeah, I meant swanctl.conf.
Thank you very much for your answers!
